plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java

```diff
@@ -57,61 +57,62 @@ public void initMLIndexIfAbsent(MLIndex index, ActionListener<Boolean> listener)
         String indexName = index.getIndexName();
         String mapping = index.getMapping();
 
-        if (!clusterService.state().metadata().hasIndex(indexName)) {
-            try (ThreadContext.StoredContext threadContext = client.threadPool().getThreadContext().stashContext()) {
+        try (ThreadContext.StoredContext threadContext = client.threadPool().getThreadContext().stashContext()) {
+            ActionListener<Boolean> internalListener = ActionListener.runBefore(listener, () -> threadContext.restore());
+            if (!clusterService.state().metadata().hasIndex(indexName)) {
```

Comment:
if we are stashing the context here, do we already know user has permission to update/create index still? I assume it maps to some write access for ml-commons.